WordPress/Classes/Networking/LoginClient.swift

```diff
@@ -78,4 +81,134 @@ actor LoginClient {
+
+        navigationController.pushViewController(loginViewController, animated: true)
+    }
+}
+
+class SelfHostedLoginViewModel: LoginWithUrlView.ViewModel {
+    private let loginClient: LoginClient
+    private let presentationContextProvider = SelfHostedLoginViewModelAuthenticationContextProvider()
+
+    var onLoginComplete: (() -> Void)?
+
+    init(loginClient: LoginClient) {
+        self.loginClient = loginClient
+        super.init()
+    }
+
+    override func startLogin() {
+        debugPrint("Attempting login with \(super.urlField)")
+        Task {
+            do {
+                let authUrl = try await self.buildLoginUrl(from: super.urlField)
+                debugPrint("Presenting Login Page \(authUrl)")
+
+                let urlWithToken = try await withUnsafeThrowingContinuation { continuation in
+                    let session = ASWebAuthenticationSession(url: authUrl, callbackURLScheme: "x-wpcom-login") { url, error in
+
+                        if let url {
+                            continuation.resume(returning: url)
+                        }
+
+                        if let error {
+                            continuation.resume(throwing: error)
+                        }
+                    }
+                    session.presentationContextProvider = self.presentationContextProvider
+                    session.start()
+                }
+
+                let credentials = try SelfHostedLoginDetails.fromApplicationPasswordResponse(urlWithToken)
+                let blogOptions = try await loadSiteOptions(details: credentials)
+
+                // Only store the new site after credentials are validated.
+                let _ = try await Blog.createRestApiBlog(with: credentials, in: ContextManager.shared)
+
+                // Automatically show the new site.
+                assert(WordPressAuthenticator.shared.delegate != nil)
+                let wporg = WordPressOrgCredentials(
+                    username: credentials.username,
+                    password: credentials.password,
+                    xmlrpc: credentials.derivedXMLRPCRoot.absoluteString,
+                    options: blogOptions
+                )
+                WordPressAuthenticator.shared.delegate!.sync(credentials: .init(wporg: wporg)) {
+                    NotificationCenter.default.post(name: Foundation.Notification.Name(rawValue: WordPressAuthenticator.WPSigninDidFinishNotification), object: nil)
+                }
+
+                self.onLoginComplete?()
+            } catch let err {
+                debugPrint(err.localizedDescription)
+            }
+        }
+    }
+
+    private func buildLoginUrl(from userInput: String) async throws -> URL {
+        let result = try await self.loginClient.performLoginAutodiscovery(for: userInput)
+        var mutableAuthURL = URL(string: result.loginUrl.url())!
+
+        var appNameValue = AppConfiguration.isJetpack ? "Jetpack iOS" : "WordPress iOS"
+
+        #if os(macOS)
```

Comment:
Is `#if os(macOS)` needed?

Reply:
Don't think so. Removed in https://github.com/wordpress-mobile/WordPress-iOS/pull/23473/commits/0fe3ea7512a61905583c99cf254e5afa1d5ed8ae